lib/ttfunk/table/cff.rb

```diff
@@ -1,27 +1,63 @@
 module TTFunk
   class Table
     class Cff < TTFunk::Table
-      autoload :Dict,   'ttfunk/table/cff/dict'
-      autoload :Header, 'ttfunk/table/cff/header'
-      autoload :Index,  'ttfunk/table/cff/index'
+      autoload :Charset,          'ttfunk/table/cff/charset'
+      autoload :Charstring,       'ttfunk/table/cff/charstring'
+      autoload :CharstringsIndex, 'ttfunk/table/cff/charstrings_index'
+      autoload :Dict,             'ttfunk/table/cff/dict'
+      autoload :Encoding,         'ttfunk/table/cff/encoding'
+      autoload :FdSelector,       'ttfunk/table/cff/fd_selector'
+      autoload :FontDict,         'ttfunk/table/cff/font_dict'
+      autoload :FontIndex,        'ttfunk/table/cff/font_index'
+      autoload :Header,           'ttfunk/table/cff/header'
+      autoload :Index,            'ttfunk/table/cff/index'
+      autoload :Path,             'ttfunk/table/cff/path'
+      autoload :PrivateDict,      'ttfunk/table/cff/private_dict'
+      autoload :SubrIndex,        'ttfunk/table/cff/subr_index'
+      autoload :TopDict,          'ttfunk/table/cff/top_dict'
+      autoload :TopIndex,         'ttfunk/table/cff/top_index'
```

Comment:
This is a very common misconception. `autoload` has been thread-safe since ruby 1.9.3. It is certainly thread-safe in ruby 2.0 and later. See: https://bugs.ruby-lang.org/issues/921

Reply:
Good news! I'm glad this issue got some attention. However it seem the feature itself is still problematic:
* Here's a relatively recent issue [#14742](https://bugs.ruby-lang.org/issues/14742). It's resolved but illustrates an old maxim of "threads are hard".
* Here's an older issue [#12688](https://bugs.ruby-lang.org/issues/12688) that haven't been resolved yet.

What I'm trying to say is that gain is marginal but potential for issues is considerable. I'd still avoid autoload if possible.

Reply:
Hmm that's definitely food for thought. However, I would say the issues you linked to are fairly edge case-y. TTFunk generally doesn't define multiple classes per file (we should fix places where it does), so the first issue where multiple constants are autoloaded from the same file shouldn't be a problem. The second issue where a class is re-opened and asked to re-autoload the same constant over and over again I would think is also a pretty rare problem.

I've been using autoload in my own libraries for years and never had problems with it. Moreover, autoload is used in bundler, rack, and sinatra, some of the most popular and used Ruby libraries.

In my opinion, the main benefit of autoload is to avoid having to require your entire library when your application (or whatever) starts up. To me it's very convenient to be able to say `require 'ttfunk'` and then be able to reference any constant within the `TTFunk` namespace. As it is, I have to specifically require `ttfunk/subset` in order to access the `TTFunk::Subset` class, which I find sort of annoying.